src/coreclr/jit/ssabuilder.cpp

```diff
@@ -853,6 +853,117 @@ void SsaBuilder::RenameDef(GenTreeOp* asgNode, BasicBlock* block)
     }
 }
 
+//------------------------------------------------------------------------
+// RenameCall: Rename a local or memory definition generated by a GT_CALL node
+//             This is only applicable for calls that takes a "hidden return buffer" argument.
+//
+// Arguments:
+//    callNode - The GT_CALL node that generates the definition
+//    block - The basic block that contains `callNode`
+//
+void SsaBuilder::RenameCall(GenTreeCall* callNode, BasicBlock* block)
```

Comment:
Re-parsing the output of `fgMorphArgs` seems very fragile.

I think it would be better to save the argument number of the hidden buffer argument (in `fgArgInfo` say -- it can be just a constant if the ABI allows it) and use it here to retrieve the local node. Unfortunately, late args will still be a complication, but that is how it is.

Also, I think it would be better to move this code into `DefinedLocal` so that we have only one routine for the front-end which can be used to answer a question about what locals does a given node define.

Reply:
We do this in other places so I would not be too worried about the assumption that the ret buf arg is the first arg after the 'this' pointer.
One thing to make it somewhat better is to add an assertion for fixed retbuf platforms, e.g.
```cpp
assert(!callNode->HasFixedRetBufArg() || (callNode->fgArgInfo->GetArgEntry(index)->nonStandardArgKind == NonStandardArgKind::FixedRetBuffer);
```

Reply:
Ah, I was a bit too quick, indeed it is not that trivial to find the ret buffer, but you can rely on all previous arguments being non-standard. So the loop can be something like:
```cpp
while (index < callNode->fgArgInfo->ArgCount())
{
  fgArgTabEntry* entry = callNode->fgArgInfo->GetArgEntry(index);
  if (entry->nonStandardArgKind == NonStandardArgKind::FixedRetBuffer || entry->nonStandardArgKind == NonStandardArgKind::RetBuffer)
  {
    // found arg buffer
    break;
  }
  index++;
}
```

Reply:
> Also, I think it would be better to move this code into DefinesLocal so that we have only one routine for the front-end which can be used to answer a question about what locals does a given node define.


Now I see the `DefinesLocal` and places where it is used, your suggestion totally makes sense. Without that I wonder why would I had to even add the condition for zero-init under BlockAssignment. I will explore this option and I am guessing that what I wanted to achieve from `fgMorphArgs` should get automatically handled but will need more work to also handle cases of `GT_CALL` alongside `GT_ASG`, e.g. [here](https://github.com/dotnet/runtime/blob/3a16a8a3c4f134c995045dd47994a5b6cd716218/src/coreclr/jit/morph.cpp#L15349-L15353). I might opt to do that as part of follow-up PR so we at least get these changes and see how they impact the benchmarks.

Reply:
I suppose I should expand on how I see things "should" be done with this change:
1) Add a routine (living on `GenTreeCall` or close to it) that will discover the hidden buffer argument post args morphing.
2) Use it in `DefinesLocal` that would then also handle `GT_CALL`. Use `DefinedLocalAddr` instead of bespoke handling for `FIELD(ADDR(LCL))`, that will also handle whether the def is partial or not correctly.
3) Delete the code from local morph and `fgMorphArgs` that sets `GTF_VAR_DEF | GTF_VAR_USEASG`, now that `DefinesLocal` handles this, `fgAssignSetVarDef` should set the flags in morph instead (it will need to be called in `fgMorphCall` after the args have been morphed).
4) Delete this `fgMorphArgs` code that is stripping the assertions, rely on the general handling (aka add `GT_CALL` to the condition).
5) Delete `RenameCall`, add appropriate handling to `RenameDef` (it should already be mostly able to deal with `GT_CALL`s as I see).
6) Use `DefinesLocal` in `fgValueNumberCall` to assign the VN to the local. Leave a `TODO` for me to unify that code with the one in `fgValueNumberAssignment` later.

Also some notes:
1) Calls with these return buffer pointers will now need to be marked as `GTF_ASG` (in `fgMorphCall` alongside `fgAssignSetVarDef`, and they have to be added to `OperRequiresAsgFlag`).
2) LIR routines (interference checking + the one Jakob added recently + possibly the debug checker) will need to be audited/updated to handle these calls.
3) Copy propagation has to be updated to push defs for these calls  -- required for correctness.
4) Notably, liveness _does not_ have to be updated, for now at least - I will switch switch it to not using location nodes at some point in the near future.
5) `optRemoveRedundantZeroInits` may need to be updated for correctness as well (not sure of this one).

Reply:
@jakobbotsch - I don't think I can rely on `NonStandardKind::FixedRetBuffer` because it is not set for non-arm64 AFAIK.

https://github.com/dotnet/runtime/blob/ee0b4607e676c72a7eefc02bc18be0e6fd7db1be/src/coreclr/jit/morph.cpp#L2513-L2523

and `hasRetBuffer` returns true only for arm64.

Reply:
May be by checking `HasRetBufArg()`.

Reply:
I am planning to move it in a method on `GenTreeCall` which will just memorize the `GenTree*` and then we can just query it instead of figuring out everytime.

Reply:
BTW, it is on my todo list to clean up the arg table and make all of this nicer. No reason we cannot have something like `NonStandardArgKind` for other "special" arguments.

Reply:
Hhm, that's true, `fgArgInfo` is not even populated during local address visitor.

Reply:
@SingleAccretion - Pushed the changes based on your suggestions. I think they are much cleaner. Thank you!